src/alpaka/AlpakaCore/HistoContainer.h

```diff
@@ -0,0 +1,370 @@
+#ifndef HeterogeneousCore_CUDAUtilities_interface_HistoContainer_h
+#define HeterogeneousCore_CUDAUtilities_interface_HistoContainer_h
+
+#include <algorithm>
+#include <cstddef>
+#include <cstdint>
+#include <type_traits>
+
+#include "AlpakaCore/alpakaConfig.h"
+#include "AlpakaCore/alpakaWorkDivHelper.h"
+#include "AlpakaCore/AtomicPairCounter.h"
+#include "AlpakaCore/alpakastdAlgorithm.h"
+#include "AlpakaCore/prefixScan.h"
+
+using namespace ALPAKA_ACCELERATOR_NAMESPACE;
+
+namespace cms {
+  namespace alpakatools {
+
+    struct countFromVector {
+      template <typename T_Acc, typename Histo, typename T>
+      ALPAKA_FN_ACC void operator()(const T_Acc &acc,
+                                    Histo *__restrict__ h,
+                                    uint32_t nh,
+                                    T const *__restrict__ v,
+                                    uint32_t const *__restrict__ offsets) const {
+        const uint32_t nt = offsets[nh];
+        const uint32_t gridDimension(alpaka::workdiv::getWorkDiv<alpaka::Grid, alpaka::Elems>(acc)[0u]);
+        const auto &[firstElementIdxNoStride, endElementIdxNoStride] =
+            cms::alpakatools::element_global_index_range(acc, Vec1::all(nt));
+        uint32_t endElementIdx = endElementIdxNoStride[0u];
+        for (uint32_t threadIdx = firstElementIdxNoStride[0u]; threadIdx < nt; threadIdx += gridDimension) {
+          for (uint32_t i = threadIdx; i < std::min(endElementIdx, nt); ++i) {
+            auto off = alpaka_std::upper_bound(offsets, offsets + nh + 1, i);
+            assert((*off) > 0);
+            int32_t ih = off - offsets - 1;
+            assert(ih >= 0);
+            assert(ih < int(nh));
+            h->count(acc, v[i], ih);
+          }
+          endElementIdx += gridDimension;
+        }
+      }
+    };
+
+    struct fillFromVector {
+      template <typename T_Acc, typename Histo, typename T>
+      ALPAKA_FN_ACC void operator()(const T_Acc &acc,
+                                    Histo *__restrict__ h,
+                                    uint32_t nh,
+                                    T const *__restrict__ v,
+                                    uint32_t const *__restrict__ offsets) const {
+        const uint32_t nt = offsets[nh];
+        const uint32_t gridDimension(alpaka::workdiv::getWorkDiv<alpaka::Grid, alpaka::Elems>(acc)[0u]);
+        const auto &[firstElementIdxNoStride, endElementIdxNoStride] =
+            cms::alpakatools::element_global_index_range(acc, Vec1::all(nt));
+
+        uint32_t endElementIdx = endElementIdxNoStride[0u];
+        for (uint32_t threadIdx = firstElementIdxNoStride[0u]; threadIdx < nt; threadIdx += gridDimension) {
+          for (uint32_t i = threadIdx; i < std::min(endElementIdx, nt); ++i) {
```

Comment:
I see this pattern
```cpp
const auto gridDim = ...;
const auto [firstNoStride, endNoStride] = ...;
for (auto threadIdx = firstNoStride, endElementIdx = endNoStride; threadIdx < MAX; threadIdx += gridDimension, endElementIdx += gridDimension) {
  for (auto i = threadIdx; i < std::min(endElementIdx, MAX); ++i ) {
    <body>
  }
}
```
repeats in almost(?) every kernel in this PR. That's 4+ lines of repetitive and error prone code. I'm wondering if it would be worth to abstract that along
```cpp
// the name should be more descriptive...
template <typename T_Acc, typename N, typename Func>
void for_each_element(const T_Acc& acc, const N nitems, Func func) {
  const auto gridDim = ...;
  const auto [firstNoStride, endNoStride] = ...;
  for (auto threadIdx = firstNoStride, endElementIdx = endNoStride; threadIdx < nitems; threadIdx += gridDimension, endElementIdx += gridDimension) {
    for (auto i = threadIdx; i < std::min(endElementIdx, nitems); ++i ) {
      func(i);
    }
  }
}
```
that could be called here along
```cpp
const uint32_t nt = offsets[nh];
for_each_element(acc, nt, [&](uint32_t i) {
  auto off = alpaka_std::upper_bound(offsets, offsets + nh + 1, i);
  ...
}
```
? I know this starts to look like we would be building our own abstraction layer on top of Alpaka, but to me the boilerplace calls for _something_. 

Written that I'm fine if this is left for a future PR.

Reply:
Yes absolutely, was thinking marking this as a to-do comment for this PR.
Maybe in an additional PR indeed is better.